tools/clang/lib/CodeGen/CGHLSLMSFinishCodeGen.cpp

```diff
@@ -2353,6 +2456,130 @@ void ProcessCtorFunctions(llvm::Module &M, StringRef globalName,
   }
 }
 
+void ProcessCtorFunctionsForLib(llvm::Module &M, llvm::StringRef globalName,
+                                DxilTypeSystem &typeSys) {
+  // Identify const static globals which only have one store in ctors first.
+  GlobalVariable *GV = M.getGlobalVariable(globalName);
+  if (!GV)
+    return;
+  ConstantArray *CA = dyn_cast<ConstantArray>(GV->getInitializer());
+  if (!CA)
+    return;
+
+  SmallVector<Function *, 2> Ctors;
+  for (User::op_iterator i = CA->op_begin(), e = CA->op_end(); i != e; ++i) {
+    if (isa<ConstantAggregateZero>(*i))
+      continue;
+    ConstantStruct *CS = cast<ConstantStruct>(*i);
+    if (isa<ConstantPointerNull>(CS->getOperand(1)))
+      continue;
+
+    // Must have a function or null ptr.
+    if (!isa<Function>(CS->getOperand(1)))
+      continue;
+    Function *Ctor = cast<Function>(CS->getOperand(1));
+    DXASSERT(Ctor->getReturnType()->isVoidTy() && Ctor->arg_size() == 0,
+             "function type must be void (void)");
+
+    for (inst_iterator I = inst_begin(Ctor), E = inst_end(Ctor); I != E; ++I) {
+      if (CallInst *CI = dyn_cast<CallInst>(&(*I))) {
+        Function *F = CI->getCalledFunction();
+        DXASSERT(F->getReturnType()->isVoidTy(), "ctor must be void(void)");
+        DXASSERT(F->arg_empty(), "ctor must be void(void)");
+        Ctors.emplace_back(F);
+      } else {
+        DXASSERT(isa<ReturnInst>(&(*I)),
+                 "else invalid Global constructor function");
+      }
+    }
+  }
+
+  for (Function *Ctor : Ctors) {
+    SmallVector<GlobalVariable *, 2> ConstStaticGV;
+    bool bHasSideEffect = false;
+    // Check all StoreInst in Ctor which is store a GV has only one store.
+    for (inst_iterator I = inst_begin(Ctor), E = inst_end(Ctor); I != E; ++I) {
+      Instruction *Inst = cast<Instruction>(&(*I));
+      StoreInst *ST = dyn_cast<StoreInst>(Inst);
+      if (!ST) {
+        if (CallInst *CI = dyn_cast<CallInst>(Inst)) {
+          if (HasSideEffect(CI, typeSys)) {
+            bHasSideEffect = true;
+            break;
+          }
+        } else if (Inst->mayHaveSideEffects()) {
+          bHasSideEffect = true;
+          break;
+        }
+        continue;
+      }
+      GlobalVariable *GV = dyn_cast<GlobalVariable>(ST->getPointerOperand());
+      if (!GV)
+        continue;
+      if (!IsSingleStoreStaticGlobal(GV, typeSys)) {
+        // Store on GV with multiple defines.
```

Comment:
As mentioned above, a false return may not mean this exactly, it may mean some other expectation was violated or case was unhandled.